tenpy/algorithms/mps_compress.py

```diff
@@ -0,0 +1,345 @@
+    """
+    theta=mps.get_theta(i)
+    theta=theta.combine_legs([['vL','p0'],['p1','vR']], qconj=[+1,-1])
+    if trunc_par==None:
+        u, s, vh, err, renorm = svd_theta(theta, {'chi_max':10000, 'svd_min':1.e-15, 'trunc_cut':1.e-15})
+        mps.norm*=renorm
+    else:
+        u, s, vh, err, renorm = svd_theta(theta, trunc_par)
+        mps.norm*=renorm
+    u=u.split_legs()
+    vh=vh.split_legs()
+    u.ireplace_label('p0', 'p')
+    vh.ireplace_label('p1', 'p')
+    mps.set_B(i, u, form='A')
+    mps.set_B((i+1)%mps.L, vh, form='B')
+    mps.set_SR(i, s)
+
+
+def apply_mpo(psi, mpo, trunc_par):
+    """Applies an mpo and truncates the resulting MPS using SVD.
+
+    Parameters
+    ----------
+    mps : :class:`~tenpy.networks.mps.MPS`
+        MPS to apply operator on
+    mpo : :class:`~tenpy.networks.mpo.MPO`
+        MPO to apply. Usually one of make_UI() or make_UII(). The approximation being made are uncontrolled for other mpos and infinite bc.
+    trunc_par : dict
+        Truncation parameters. See :func:`truncate`
+
+    Returns
+    -------
+    mps : :class:`~tenpy.networks.mps.MPS` 
+        Resulting new MPS
+    """
+    bc=psi.bc
+    if bc!=mpo.bc:
+        raise ValueError("Boundary conditions of MPS and MPO are not the same")
+    if psi.L != mpo.L:
+        raise ValueError("Length of MPS and MPO not the same")
+    Bs=[npc.tensordot(psi.get_B(i, form='B'), mpo.get_W(i), axes=('p', 'p*')) for i in range(psi.L)]
+    Bs[0]=npc.tensordot(psi.get_theta(0,1), mpo.get_W(0), axes=('p0', 'p*'))
+    for i in range(psi.L):
+        if i==0 and bc=='finite':
+            Bs[i]=Bs[i].take_slice(mpo.get_IdL(i) ,'wL')
+            Bs[i]=Bs[i].combine_legs(['wR','vR'], qconj=[-1])
+            Bs[i].ireplace_labels(['(wR.vR)'], ['vR'])
+            Bs[i].get_leg('vR').to_LegCharge
+        elif i==psi.L-1 and bc=='finite':
+            Bs[i]=Bs[i].take_slice(mpo.get_IdR(i) ,'wR')
+            Bs[i]=Bs[i].combine_legs(['wL','vL'], qconj=[1])
+            Bs[i].ireplace_labels(['(wL.vL)'], ['vL'])
+            Bs[i].get_leg('vL').to_LegCharge()
+        else:
+            Bs[i]=Bs[i].combine_legs([['wL','vL'],['wR','vR']], qconj=[+1, -1])
+            Bs[i].ireplace_labels(['(wL.vL)', '(wR.vR)'], ['vL','vR'])
+            Bs[i].get_leg('vL').to_LegCharge()
+            Bs[i].get_leg('vR').to_LegCharge()
+
+    if bc=='infinite':
+        #calculate good (rather arbitrary) guess for S[0] (no we don't like it either)
+        weight=np.ones(mpo.get_W(0).shape[mpo.get_W(0).get_leg_index('wL')])*0.05
+        weight[mpo.get_IdL(0)]=1
+        weight=weight/np.linalg.norm(weight)
+        S=[np.kron(psi.get_SL(0), weight)]
+    else:
+        S=[np.ones(Bs[0].shape[Bs[0].get_leg_index('vL')])]
+    #Wrong S values but will be calculated in mps_compress
+    for i in range(psi.L):
+        S.append(np.ones(Bs[i].shape[Bs[i].get_leg_index('vR')]))

```

Comment:
`Bs[i].get_leg('vR').ind_len` is equivalent but shorter than `Bs[i].shape[Bs[i].get_leg_index('vR')]`
(Also 3 lines above for `'vL'` leg).